coil-network/src/commonMain/kotlin/coil3/fetch/NetworkFetcher.kt

```diff
@@ -0,0 +1,289 @@
+            throw e
+        }
+    }
+
+    private fun newRequest(): HttpRequestBuilder {
+        val request = HttpRequestBuilder()
+        request.method = options.httpMethod
+        request.url.takeFrom(url)
+        request.headers.appendAll(options.httpHeaders)
+
+        val diskRead = options.diskCachePolicy.readEnabled
+        val networkRead = options.networkCachePolicy.readEnabled
+        when {
+            !networkRead && diskRead -> {
+                request.header(CACHE_CONTROL, "only-if-cached, max-stale=2147483647")
+            }
+
+            networkRead && !diskRead -> if (options.diskCachePolicy.writeEnabled) {
+                request.header(CACHE_CONTROL, "no-cache")
+            } else {
+                request.header(CACHE_CONTROL, "no-cache, no-store")
+            }
+
+            !networkRead && !diskRead -> {
+                // This causes the request to fail with a 504 Unsatisfiable Request.
+                request.header(CACHE_CONTROL, "no-cache, only-if-cached")
+            }
+        }
+
+        return request
+    }
+
+    private suspend fun <T> executeNetworkRequest(
+        request: HttpRequestBuilder,
+        block: suspend (HttpResponse) -> T,
+    ): T {
+        // Prevent executing requests on the main thread that could block due to a
+        // networking operation.
+        if (options.networkCachePolicy.readEnabled) {
+            assertNotOnMainThread()
+        }
+
+        return httpClient.value.prepareRequest(request).execute { response ->
+            if (!response.status.isSuccess() && response.status != HttpStatusCode.NotModified) {
+                throw HttpException(response)
+            }
+            block(response)
+        }
+    }
+
+    /**
+     * Parse the response's `content-type` header.
+     *
+     * "text/plain" is often used as a default/fallback MIME type.
+     * Attempt to guess a better MIME type from the file extension.
+     */
+    internal fun getMimeType(url: String, contentType: String?): String? {
+        if (contentType == null || contentType.startsWith(MIME_TYPE_TEXT_PLAIN)) {
+            MimeTypeMap.getMimeTypeFromUrl(url)?.let { return it }
+        }
+        return contentType?.substringBefore(';')
+    }
+
+    private fun DiskCache.Snapshot.toCacheResponse(): CacheResponse? {
+        try {
+            return fileSystem.read(metadata) {
+                CacheResponse(this)
+            }
+        } catch (_: IOException) {
+            // If we can't parse the metadata, ignore this entry.
+            return null
+        }
+    }
+
+    private fun DiskCache.Snapshot.toImageSource(): ImageSource {
+        return ImageSource(
+            file = data,
+            fileSystem = fileSystem,
+            diskCacheKey = diskCacheKey,
+            closeable = this,
+        )
+    }
+
+    private suspend fun ByteReadChannel.toImageSource(): ImageSource {
+        return ImageSource(
+            source = Buffer().apply { writeTo(this) },
```

Comment:
This will squash the whole http body to art heap if no diskcache available
May cause OOM if image blob too large

Reply:
Yep, this is only used as a fallback if we can't write to disk. Skiko also requires an image to be fully loaded into memory before decoding: https://github.com/JetBrains/skiko/issues/741

Reply:
It seems like this is the only way to do it.
Impossible to wrap a ByteReadChannel with okio Source(or InputStream), unless we pay for a thread blocking on it, as okio is sync and ktor is async,